src/RequestProcessor.cs

```diff
@@ -38,7 +37,6 @@ internal RequestProcessor(MessagingStream msgStream)
         {
```

Comment:
In the we're still awaiting for the rpc messages to be complete, and in this case the host doesn't send a new message to the client until it receives the response, so, for example if a function1 invocation request arrives while dependency download is going on then on azure portal user will see the message saying dependency download is in-progress, but if another function2 invocation request is received then host won't send the second invocation request to PowerShell language worker, and user won't see managed dependency download message on the azure portal, is it possible for PowerShell worker to receive and process invocation request concurrently?

 internal async Task ProcessRequestLoop()

        internal async Task ProcessRequestLoop()

        {

        {

            StreamingMessage request, response;

            StreamingMessage request, response;

            while (await _msgStream.MoveNext())

Reply:
Yes, with the current implementation, the second invocation request will be blocked until the dependency download finishes, even if the user set `PSWorkerInProcConcurrencyUpperBound` greater than 1.
It can be addressed by not blocking the main thread but create a task to call `ProcessInvocationRequest` for each `InvocationRequest` message. The was one of my designs back when I worked on the in-proc concurrency, but that design resulted in heavy lock contention and I didn't use it in the end. See the **Design Change** section in the PR description of https://github.com/Azure/azure-functions-powershell-worker/pull/123. Quoted here:

> The original design is to not block the main thread -- it reads a request message, start a task for processing and forget about it, then it reads the next message.
With this design, every in-coming request will be handled immediately -- a task will be created for it. So, a large amount of tasks will be created rapidly, and thus a large amount of thread-pool threads will be assigned to run those tasks. All those threads will compete for an idle PowerShellManager from the pool, which results in lock contention in the pool. The average time for fetching an idle PowerShellManager instance gets worse dramatically when velocity of in-coming requests reaches about 70/s -- it reaches 994 ms in my measurements, and becomes a huge bottleneck.

I guess we can revisit this design and consider the `Rx.net`, but the competition on acquiring a `PowerShellManager` instance from the pool will continue to be the source of lock contention.

Or, think it out of the box, what if we can change the design of dependency management? Separate it from the worker and make it a service or something and can be reused by all languages (e.g. the retry logic shouldn't be implemented once per worker, right?), and make the portal more aware of the dependency management component. That would be the more long term solution to all the issues to the dependency management feature.

Reply:
This is definitely something worth more discussion, but I think it's out scope for this PR.
@SatishRanjan can you please create a more general issue for the challenges of the current dependency management design? @TylerLeonhardt and I talked about this a couple times, and we both think separating the dependency management logic from the worker would be the long term solution and we should start moving toward that.

Reply:
@daxian-dbw - currently since there's await _msgStream.MoveNext() to move on next message from host to worker hence host does not even send invocation request to the powershell worker runtime, and I was thinking if we can remove wait on _msgStream.MoveNext() and concurrently process the invocation request, so if we do this way then the second invocation request will get dependency download in progress message as well?

Reply:
@SatishRanjan I don't think we should discuss the design/fix for another issue within this PR review. As I mentioned above, I suggest you to open a new issue and discussion can happen there.

Reply:
@daxian-dbw => I was expecting adding capability to process multiple invocation request concurrently was the scope of this throughput improvement change. But yes, if we want to take it separately then that should be fine as well.

Reply:
@SatishRanjan I think you misunderstood the issue your brought up here.

- First, the capability to process multiple invocations concurrently has been added to the PowerShell worker for quite a while (in Jan.), see #123. Processing multiple invocations concurrently works fine after the dependency downloading finishes. The concurrency level is controlled by the env variable `PSWorkerInProcConcurrencyUpperBound` which the user can set in the application settings.

- Second, what you brought up here is a very specific scenario -- multiple invocation triggered while the dependency downloading is in progress. By definition, no invocation can actually be served when the downloading is still in progress, so **what you really intend to address here** is to make all invocations receive the "downloading in progress" logging message.

- Third, also the most important one, the throughput should always be measured when the application is running in steady state, which means the cold start, where the downloading happens, is not in consideration when talking about throughput. Please take a look at the `Testing environment` section in my PR description, which I made it very clear that each measure is done after warming up the host/worker (`WCAT` will do another 30 seconds warming up).